api/src/extensions/lib/sync-extensions.ts

```diff
@@ -0,0 +1,94 @@
+import { NESTED_EXTENSION_TYPES } from '@directus/extensions';
+import { ensureExtensionDirs } from '@directus/extensions/node';
+import mid from 'node-machine-id';
+import { createWriteStream } from 'node:fs';
+import { mkdir } from 'node:fs/promises';
+import { uptime } from 'node:os';
+import { dirname, join, normalize, resolve, sep } from 'node:path';
+import { pipeline } from 'node:stream/promises';
+import Queue from 'p-queue';
+import env from '../../env.js';
+import logger from '../../logger.js';
+import { getMessenger } from '../../messenger.js';
+import { getStorage } from '../../storage/index.js';
+import { getExtensionsPath } from './get-extensions-path.js';
+import { SyncStatus, getSyncStatus, setSyncStatus } from './sync-status.js';
+import { getSyncTimestamp, setSyncTimestamp } from './sync-timestamp.js';
+
+export const syncExtensions = async () => {
+	const messenger = getMessenger();
+
+	const isPrimaryProcess =
+		String(process.env['NODE_APP_INSTANCE']) === '0' || process.env['NODE_APP_INSTANCE'] === undefined;
+
+	const isDone = (await getSyncStatus()) === SyncStatus.DONE;
+
+	const id = await mid.machineId();
+
+	const message = `extensions-sync/${id}`;
+
+	if (isDone === false && isPrimaryProcess === false) {
+		logger.trace('Extensions already being synced to this machine from another process.');
+
+		/**
+		 * Wait until the process that called the lock publishes a message that the syncing is complete
+		 */
+		return new Promise((resolve) => {
+			messenger.subscribe(message, resolve);
+		});
+	}
+
+	if (!env['EXTENSIONS_LOCATION']) {
+		return;
+	}
```

Comment:
Shouldn't we move this check to the top?

Reply:
Yeah 👍🏻 Remnant of when this file was shared with other extension loading biz